src/Maps/MapWriter.cpp

```diff
@@ -46,13 +48,18 @@ namespace MapWriter {
 
 
 	namespace {
-		void WriteHeader(Stream::Writer& streamWriter, const MapHeader& header)
-		{
-			if (!header.VersionTagValid()) {
-				throw std::runtime_error("All instances of version tag in .map and .op2 files must be greater than 0x1010.");
+		void ValidateMap(const MapData& mapData) {
+			if (!mapData.header.VersionTagValid()) {
+				throw std::runtime_error("All instances of version tag in .map and .op2 files must be greater than 0x1010");
 			}
 
-			streamWriter.Write(header);
+			if (mapData.header.TileCount() != mapData.tiles.size()) {
+				throw std::runtime_error("Header reported tile width * tile height does not match actual tile count");
+			}
+
+			if (mapData.header.numTilesets != mapData.tilesetSources.size()) {
+				throw std::runtime_error("Header reported tileset count does not match actual tileset sources");
+			}
```

Comment:
I'm not convinced we need the `ValidateMap` method. I think perhaps the `header` struct could be synthesized from the other fields, rather than validated against them. That could greatly simplify the need for error checking. It also means we're not pushing the problem onto the user of the library, where they have to manually fill in the `header` from all the component parts.